src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -321,87 +316,49 @@ namespace xclhwemhal2 {
     auto top = reinterpret_cast<const axlf*>(header);
     if (auto sec = xclbin::get_axlf_section(top, EMBEDDED_METADATA)) {
       xmlFileSize = sec->m_sectionSize;
-      xmlFile = new char[xmlFileSize];
-      memcpy(xmlFile, bitstreambin + sec->m_sectionOffset, xmlFileSize);
+      xmlFile = std::make_unique<char[]>(xmlFileSize);
+      memcpy(xmlFile.get(), bitstreambin + sec->m_sectionOffset, xmlFileSize);
     }
     if (auto sec = xclbin::get_axlf_section(top, BITSTREAM)) {
       zipFileSize = sec->m_sectionSize;
-      zipFile = new char[zipFileSize];
-      memcpy(zipFile, bitstreambin + sec->m_sectionOffset, zipFileSize);
+      zipFile = std::make_unique<char[]>(zipFileSize);
+      memcpy(zipFile.get(), bitstreambin + sec->m_sectionOffset, zipFileSize);
     }
     if (auto sec = xclbin::get_axlf_section(top, DEBUG_IP_LAYOUT)) {
       debugFileSize = sec->m_sectionSize;
-      debugFile = new char[debugFileSize];
-      memcpy(debugFile, bitstreambin + sec->m_sectionOffset, debugFileSize);
+      debugFile = std::make_unique<char[]>(debugFileSize);
+      memcpy(debugFile.get(), bitstreambin + sec->m_sectionOffset, debugFileSize);
     }
     if (auto sec = xrt_core::xclbin::get_axlf_section(top, ASK_GROUP_TOPOLOGY)) {
       memTopologySize = sec->m_sectionSize;
-      memTopology = new char[memTopologySize];
-      memcpy(memTopology, bitstreambin + sec->m_sectionOffset, memTopologySize);
+      memTopology = std::make_unique<char[]>(memTopologySize);
+      memcpy(memTopology.get(), bitstreambin + sec->m_sectionOffset, memTopologySize);
     }
     if (auto sec = xclbin::get_axlf_section(top, PDI)) {
       pdiSize = sec->m_sectionSize;
-      pdi = new char[pdiSize];
-      memcpy(pdi, bitstreambin + sec->m_sectionOffset, pdiSize);
+      pdi = std::make_unique<char[]>(pdiSize);
+      memcpy(pdi.get(), bitstreambin + sec->m_sectionOffset, pdiSize);
     }
     if (auto sec = xclbin::get_axlf_section(top, EMULATION_DATA)) {
       emuDataSize = sec->m_sectionSize;
-      emuData = new char[emuDataSize];
-      memcpy(emuData, bitstreambin + sec->m_sectionOffset, emuDataSize);
-    }
-
-    if(!zipFile || !xmlFile)
-    {
-      //deallocate all allocated memories to fix memory leak
-      if(zipFile)
-      {
-        delete[] zipFile;
-        zipFile = nullptr;
-      }
-
-      if(debugFile)
-      {
-        delete[] debugFile;
-        debugFile = nullptr;
-      }
-
-      if(xmlFile)
-      {
-        delete[] xmlFile;
-        xmlFile = nullptr;
-      }
-
-      if(memTopology)
-      {
-        delete[] memTopology;
-        memTopology = nullptr;
-      }
-
-      if (pdi) {
-        delete[] pdi;
-        pdi = nullptr;
-      }
-
-      if (emuData) {
-        delete[] emuData;
-        emuData = nullptr;
-      }
-      return -1;
+      emuData = std::make_unique<char[]>(emuDataSize);
+      memcpy(emuData.get(), bitstreambin + sec->m_sectionOffset, emuDataSize);
```

Comment:
While looking at this assembly code, I have the feeling that `std::vector<char>` everywhere would be simpler. But probably an über-refactoring for next century.

Reply:
@keryell I was thinking the same, but didn't bother to comment on it.  Maybe with the exception of where you need a custom deleter, I don't think std::unique_ptr<[]> should be used.

Reply:
@stsoe So, **instead of** "unique_ptr" should i use "new" then ... Please confirm !!

Reply:
@udaybhas-xilinx You should never ever use 'new and delete'.
What we are saying is that `std::vector<char> vec(size)` is as good as, or better than, `std::unique_ptr<char[]>`